mithril-common/src/cardano_transaction_parser.rs

```diff
@@ -119,48 +115,60 @@ impl CardanoTransactionParser {
-                    Outcome::Matched(_) | Outcome::EpochBoundary => true,
-                    Outcome::Inconclusive => false,
-                }
-            })
-            .collect();
-
-        blocks_start_byte_index.push(cbor.len() + 1);
+        let hardano_blocks = hardano_blocks(immutable_file)?;
 
         let mut blocks = Vec::new();
-        let mut last_start_byte_index = 0;
-        for block_start_index in blocks_start_byte_index.into_iter().skip(1) {
-            let maybe_end_byte_index = min(block_start_index, cbor.len());
-            if let Ok(multi_era_block) =
-                MultiEraBlock::decode(&cbor[last_start_byte_index..maybe_end_byte_index])
-            {
-                let block = Block::try_convert(multi_era_block, immutable_file.number)
-                    .with_context(|| {
-                        format!(
-                        "CardanoTransactionParser could not read data from block in immutable file: {:?}",
+        for block in hardano_blocks {
+            let block = block.map_err(|e| {
+                anyhow!(e).context(format!(
+                    "Error while reading block in immutable file: '{:?}'",
+                    immutable_file.path
+                ))
+            })?;
+
+            match MultiEraBlock::decode(&block) {
+                Ok(multi_era_block) => {
+                    let block = Block::try_convert(multi_era_block, immutable_file.number)
+                        .with_context(|| {
+                            format!(
+                            "CardanoTransactionParser could not read data from block in immutable file: {:?}",
+                            immutable_file.path
+                        )
+                        })?;
+                    blocks.push(block);
+                }
+                Err(err) => {
+                    return Err(anyhow!(err).context(format!(
+                        "Error while decoding block in immutable file: '{:?}'",
                         immutable_file.path
-                    )
-                    })?;
-                blocks.push(block);
-                last_start_byte_index = block_start_index;
+                    )))
+                }
             }
         }
 
         Ok(blocks)
     }
 }
 
+fn hardano_blocks(
+    immutable_file: &ImmutableFile,
+) -> StdResult<pallas_hardano::storage::immutable::chunk::Reader> {
+    let dir_path = immutable_file.path.parent().ok_or(anyhow!(format!(
+        "Could not retrieve immutable file directory with immutable file path: '{:?}'",
+        immutable_file.path
+    )))?;
+    let file_name = &Path::new(&immutable_file.filename)
+        .file_stem()
+        .ok_or(anyhow!(format!(
+            "Could not extract immutable file name from file: '{}'",
+            immutable_file.filename
+        )))?
+        .to_string_lossy();
+    let blocks = pallas_hardano::storage::immutable::chunk::read_blocks(dir_path, file_name)?;
```

Comment:
This should be imported at the beginning of the module: `pallas_hardano::storage::immutable::chunk`. 
The comment also applies to all other imports.